2023-Oct-18/Dave/word_grid.py

```diff
@@ -0,0 +1,76 @@
+import random
+import string
+from point import Point
+from direction import Direction
+
+
+class WordGrid:
+    def __init__(self, rows: int, cols: int):
+        self.rows = rows
+        self.cols = cols
+        self.initialise_word_grid()
+
+    def __repr__(self):
+        return self.get_stringified_word_grid()
+
+    def __iter__(self):
+        for row in self.word_grid:
+            yield row
+
+    def initialise_word_grid(self):
+        self.word_grid = [[" " for col in range(
+            self.cols)] for row in range(self.rows)]
+
+    def get_stringified_word_grid(self) -> str:
+        output = ""
+        for row in self.word_grid:
+            for char in row:
+                output += char + " "
+            output += "\n"
+        return output
+
+    def read_line(
+        self, position: Point, orientation: Direction, length: int
+    ) -> str | None:
+        result = ""
+
+        current_row, current_col = position
+        next_row, next_col = orientation
+        grid_rows = len(self.word_grid)
+        grid_cols = len(self.word_grid[0])
+
+        for i in range(length):
+            if not 0 <= current_row < len(self.word_grid):
+                return None
+
+            if not 0 <= current_col < len(self.word_grid[0]):
+                return None
+
+            result += self.word_grid[current_row][current_col]
+            current_row += next_row
+            current_col += next_col
+
+        return result
+
+    def write_line(self, position: Point, orientation: Direction, data: str) -> bool:
+        can_write_line = self.read_line(position, orientation, len(data))
+        if not can_write_line:
+            return False
+
+        current_row, current_col = position
+        next_row, next_col = orientation
+
+        for char in data:
+            self.word_grid[current_row][current_col] = char
+            current_row += next_row
+            current_col += next_col
+
+        return True
+
+    def fill_blank_space(self):
+        for row_index, row in enumerate(self.word_grid):
+            for col_index, char in enumerate(row):
+                if char == " ":
+                    self.word_grid[row_index][col_index] = random.choice(
+                        string.ascii_uppercase
+                    )
```

Comment:
Your could fill the grid with `None`s instead of chars as this should be quicker for python to check, but then you have two types in a container which I am not a fan of either so up to you. 

I would suggest using `' '` for characters as this is a restriction in other languages ( like C/C++ and Java)

I like the `string.ascii_uppercase`, did not know you could do that!

Reply:
I think the space is possibly preferable here, as the empty cells can be concatenated into strings to produce the return from read_line and None won't concatenate in that way.

Also my linter presently switches single-quotes for double-quotes, so I'm not gonna fight with it until the language complains :)

Reply:
Yes you are right! I assumed they were a container of some kind!